S3 select

sinhaashish opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
 SQL select query for CSV/JSON is implemented with a generic code.
## Description
<!--- Describe your changes in detail -->
select * , select column names has been implemented for CSV.
select * is implemented fot JSON. Code is made generic.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
S3  select for JSON and CSV is implemented
<!--- If it fixes an open issue, please link to the issue here. -->

## Regression
<!-- Is this PR fixing a regression? (Yes / No) -->
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
Select * , select columns are working for CSV.
Select * is working for JSON.

tested with python 
* For CSV 
```from botocore.client import Config
import os
import boto3
s3 = boto3.resource('s3',
                    endpoint_url='http://127.0.0.1:9000',
                    aws_access_key_id='minio',
                    aws_secret_access_key='minio123',
                    config=Config(signature_version='s3v4'),
                    region_name='us-east-1')
s3_client = s3.meta.client
r = s3_client.select_object_content(
        Bucket='sinha',
        Key='player.csv',
        ExpressionType='SQL',
	Expression = "select * from S3Object AS S  where _3 ='Acy'  ",
      #Expression = "select * from S3Object AS S  where last_name='Acy'  ",
       InputSerialization = {'CSV': {"FieldDelimiter": ",","FileHeaderInfo":"NONE"}},
       #InputSerialization = {'CSV': {"FieldDelimiter": ",","FileHeaderInfo":"USE"}},
        OutputSerialization = {'CSV': {}},
    )

event_stream = r['Payload']
end_event_received = False
with open('output', 'wb') as f:
    # Iterate over events in the event stream as they come
    for event in event_stream:
        # If we received a records event, write the data to a file
        if 'Records' in event:
            data = event['Records']['Payload']
            f.write(data)
        # If we received a progress event, print the details
        elif 'Progress' in event:
            print(event['Progress']['Details'])
        # End event indicates that the request finished successfully
        elif 'End' in event:
            print('Result is complete')
            end_event_received = True
if not end_event_received:
    raise Exception("End event not received, request incomplete.")
```
* Tested with JSON
```
{ "name" : "q", "color" : "blue" }
{ "name" : "w", "color" : "red"}
{ "name" : "e", "color" : "green"}
{ "name" : "r", "color" : "orange"}
``` 

* For JSON
```from botocore.client import Config
import os
import boto3
s3 = boto3.resource('s3',
                    endpoint_url='http://127.0.0.1:9000',
                    aws_access_key_id='minio',
                    aws_secret_access_key='minio123',
                    config=Config(signature_version='s3v4'),
                    region_name='us-east-1')
s3_client = s3.meta.client

	
r = s3_client.select_object_content(
		Bucket='sinha',
		Key='AmazonExample.json',
		ExpressionType='SQL',
		Expression="select * from s3object s where s.name = 'q' limit 5 ",
		InputSerialization = { 'CompressionType': 'NONE','JSON': {'Type': 'DOCUMENT'}},
		OutputSerialization = {'JSON': { 'RecordDelimiter': '\n',}}
)
	


event_stream = r['Payload']
end_event_received = False
with open('output', 'wb') as f:
    # Iterate over events in the event stream as they come
    for event in event_stream:
        # If we received a records event, write the data to a file
        if 'Records' in event:
            data = event['Records']['Payload']
            f.write(data)
        # If we received a progress event, print the details
        elif 'Progress' in event:
            print(event['Progress']['Details'])
        # End event indicates that the request finished successfully
        elif 'End' in event:
            print('Result is complete')
            end_event_received = True
if not end_event_received:
    raise Exception("End event not received, request incomplete.")
```

<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

kannappanr commented on pkg/s3select/select-datatypes.go:
```diff
@@ -0,0 +1,105 @@
+/*
+ * Minio Cloud Storage, (C) 2015, 2016 Minio, Inc.
+ *
```
Copyright should be 2018.

kannappanr commented on vendor/github.com/json-iterator/go/Gopkg.lock:
```diff
@@ -0,0 +1,21 @@
+# This file is autogenerated, do not edit; changes may be undone by the next 'dep ensure'.
```
Do we need `Gopkg.lock` and `Gopkg.toml` files? I know these are generated using `dep ensure`. Either way is fine with me.

kannappanr commented on cmd/object-handlers.go:
```diff
@@ -165,22 +165,22 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
 		selectReq.InputSerialization.CompressionType == "" {
-		selectReq.InputSerialization.CompressionType = SelectCompressionNONE
-		if !strings.Contains(objInfo.ContentType, "text/csv") {
+		selectReq.InputSerialization.CompressionType = s3select.SelectCompressionNONE
```
The first part of the `if` check can be removed. The value is being checked for and set to the same value again.

sinhaashish replied:
Done.

kannappanr commented on cmd/object-handlers.go:
```diff
@@ -193,26 +193,39 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
-		writeErrorResponse(w, ErrInvalidQuoteFields, r.URL)
-		return
+
+	if selectReq.InputSerialization.CSV != nil {
```
`if` checks in Lines 197 and 198 are contradicting each other. I am guessing you are doing this to incorporate checks for `json` also. But this is wrong. It can be something like, if all 4 parameters (Inputserialization.CSV, OutputSerialization.CSV, InputSerialization.JSON and OutputSerialization.JSON) are all empty, then you need to return error. If one of them (Say InputSerializaion.CSV) is set and the other one (OututSerialization.CSV) is not set, then you need to fail. Same test needs to be done for JSON. Can a user send a query to both CSV and JSON objects in one call?

sinhaashish replied:
Done

kannappanr commented on cmd/object-handlers.go:
```diff
@@ -248,22 +263,41 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
-			logger.LogIf(ctx, err)
-		}
+	}
+	_, _, _, _, _, _, err = s3select.ParseSelect(selectReq.Expression, s3s)
```
Are the return values of `ParseSelect` used anywhere if not, you can just return `err`

sinhaashish replied:
In ```object-handlers.go``` only the error is checked  but in ```select.go``` all the returned values are used.

kannappanr replied:
ok

kannappanr commented on pkg/s3select/helpers.go:
```diff
@@ -70,42 +72,28 @@ func representsInt(s string) bool {
 	}
+	out, err := json.Marshal(individualRecord)
+	if err != nil {
+		fmt.Println("\n  error:", err)
```
Please remove `fmt.Println` and introduce LogIf, if it needs to be logged. Shouldn't you return in the case of failure?

kannappanr replied:
Also, why is `json.Marshal` introduced here, can you please add a comment on why it was not there before and what it is doing now.

kannappanr commented on pkg/s3select/helpers.go:
```diff
@@ -70,42 +72,28 @@ func representsInt(s string) bool {
-			return false, err
-		}
-		return myResult, nil
+		//TODO
```
Why is this removed? and what does this affect by this removal.

sinhaashish replied:
It is for ```BETWEEN``` clause . You cant run the where clause with ```BETWEEN``` keyword unless it is implemented.

kannappanr commented on pkg/s3select/input.go:
```diff
@@ -75,6 +80,143 @@ type Input struct {
+	// Do it for now but we have to take the value not the keys for CSV
+	out, err := json.Marshal(individualRecord)
+	if err != nil {
+		fmt.Println("error:", err)
```
please remove all `println` s

kannappanr replied:
Also, it will be cleaner to read if you can separate out the implementations of both CSV and JSON and not interleave them. It is OK to have two implementations in separate files too.

sinhaashish replied:
```println``` removed . For Json and CSV file , we can have it as different files .Will do it in another PR

kannappanr commented on pkg/s3select/input.go:
```diff
@@ -109,13 +251,46 @@ type Options struct {
 	// Size of incoming object
 	StreamSize int64
 
+	// OutputSerializationValue
```
Do you plan to add something for later

sinhaashish replied:
OutputSerializationValue will be used to tell the Output type i.e. csv / Json

kannappanr replied:
meaning that the file can be `CSV` or `JSON` and we will still allow the output to be of certain type? does S3 allow this behavior?

sinhaashish replied:
S3 docs says ```The format of the InputSerialization doesn't need to match the format of the OutputSerialization. So, for example, you can specify JSON in the InputSerialization and CSV in the OutputSerialization.```

kannappanr replied:
ok

kannappanr reviewed (changes requested):
Please fix/respond to the coments

harshavardhana commented on pkg/s3select/funcEval.go:
```diff
@@ -186,26 +184,50 @@ func (reader *Input) evaluateFuncErr(myVal *sqlparser.FuncExpr) error {
 	// case for nested func val
 	case *sqlparser.FuncExpr:
-		myVal = evaluateFuncExpr(myIs, "", row, columnNames)
+		//myVal = evaluateFuncExpr(myIs, "", row, columnNames)
```
These type of comments should be avoided before the PR is ready for review @sinhaashish, can you make sure to avoid these?

sinhaashish replied:
From next time will cross check before raising PR

harshavardhana commented on pkg/s3select/helpers.go:
```diff
@@ -70,42 +72,28 @@ func representsInt(s string) bool {
+func matchesMyWhereClause(individualRecord map[string]interface{}, alias string, whereClause interface{}) (bool, error) {
 	var conversionColumn string
-	var operator string
+	var operator string //It gives the right value of where clause
```
Comments should be `//<space>` native Go style

sinhaashish replied:
Corrected.

harshavardhana commented on pkg/s3select/input.go:
```diff
@@ -109,13 +258,46 @@ type Options struct {
 	StreamSize int64
 
+	// OutputSerializationValue
+	//OutputSerial
```
what is this for?

sinhaashish replied:
removed.

harshavardhana commented on pkg/s3select/input.go:
```diff
@@ -288,20 +532,32 @@ type Row struct {
+	case *JSONInput:
+		expression = inputType.getExpression()
+
+	case *Input:
```
should this be named as `CSVInput` ?

sinhaashish replied:
Changed Input to CSVInput.

harshavardhana commented on pkg/s3select/input.go:
```diff
@@ -324,17 +581,19 @@ func (reader *Input) Execute(writer io.Writer) error {
 				curBuf.Reset()
-				reader.stats.BytesReturned += int64(len(row.record))
+				inputType.bytesReturned(int64(len(row.record)))
+				//reader.stats.BytesReturned = reader.stats.BytesReturned + int64(len(row.record))
```
why is this commented?

sinhaashish replied:
Modified

harshavardhana commented on pkg/s3select/json-funcEval.go:
```diff
@@ -0,0 +1,51 @@
+		myVal = string(myIs.Val)
+	// case for nested func val
+	case *sqlparser.FuncExpr:
+		//myVal = evaluateFuncExpr(myIs, "", row, columnNames)
```
why is this commented?

sinhaashish replied:
Removed the file as it was JSON specific.

harshavardhana commented on pkg/s3select/select-datatypes.go:
```diff
@@ -0,0 +1,105 @@
+/*
+ * Minio Cloud Storage, (C) 2015, 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
We can just say this as datatypes.go not `select-datatypes.go`

sinhaashish replied:
Done.

harshavardhana commented on pkg/s3select/select-datatypes.go:
```diff
@@ -0,0 +1,105 @@
+/*
+ * Minio Cloud Storage, (C) 2015, 2018 Minio, Inc.
```
Remove 2015, this is a new file only needs 2018

sinhaashish replied:
Modified

kannappanr commented on pkg/s3select/select_test.go:
```diff
@@ -116,32 +117,34 @@ func TestMyRowIndexResults(t *testing.T) {
-		{[]string{"3", "2"}, make(map[string]int), make(map[string]bool), make(map[string]int), options, s3s, []string{"random", "hullo", "thing", "stuff"}, "thing,hullo", "random,hullo,thing,stuff", []string{"1", "2", "3", "4"}, nil},
-		{[]string{"11", "1"}, make(map[string]int), make(map[string]bool), make(map[string]int), options, s3s, []string{"random", "hullo", "thing", "stuff"}, "", "random,hullo,thing,stuff", []string{"1", "2", "3", "4"}, ErrInvalidColumnIndex},
+		{[]string{"1", "2"}, make(map[string]int), make(map[string]bool), make(map[string]int), options, s3s, "'1':'random','2':'hullo','3':'thing','4':'stuff'", "target,random", "target,random,hello,stuff", []string{"1", "2", "3", "4"}, "", nil},
+		// 	{[]string{"2", "3", "4"}, make(map[string]int), make(map[string]bool), make(map[string]int), options, s3s, []string{"random", "hullo", "thing", "stuff"}, "hullo,thing,stuff", "random,hullo,thing,stuff", []string{"1", "2", "3", "", "4"}, "", nil},
```
Is there are a reason to keep commented code here?

sinhaashish replied:
The aggregate functions is already present. Have removed the commented code .

kannappanr commented on pkg/s3select/select_test.go:
```diff
@@ -116,32 +117,34 @@ func TestMyRowIndexResults(t *testing.T) {
-			t.Error()
+			//t.Error(),"random", "hello", "stuff"}
 		}
+
```
Please remove commented code

kannappanr commented on pkg/s3select/select_test.go:
```diff
@@ -360,19 +364,22 @@ func TestToStringAgg(t *testing.T) {
-		{[]float64{10, 11, 12, 13, 14}, "10.000000,11.000000,12.000000,13.000000,14.000000"},
-		{[]float64{10}, "10.000000"},
+		{[]float64{10, 11, 12, 13, 14}, "10,11,12,13,14"},
+		// {[]float64{10, 11.3, 12, 13, 14}, "10,11.300000,12,13,14"},
```
please remove commented code

sinhaashish replied:
done

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -199,17 +207,20 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 			close(myRow)
 			return
 		}
-		// The call to the where function clause,ensures that the rows we print match our where clause.
```
why was this comment removed?

sinhaashish replied:
Added the comment back

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -220,25 +231,35 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
-					// This checks whether any aggregation function was called as now we
-					// no longer will go through printing each row, and only print at the
-					// end
+					//This checks whether any aggregation function was called as now we
```
Comments need to start after a space

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 	myRow := record[0]
 	for i := 1; i < len(record); i++ {
-		myRow = myRow + reader.options.OutputFieldDelimiter + record[i]
+		myRow = myRow + inputType.getOutPutFieldDelimitter() + record[i]
```
typo `getOutPutFieldDelimitter` should be `getOutputFieldDelimitter`

sinhaashish replied:
Done

kannappanr replied:
Sorry, missed one more type `getOutputFieldDelimitter` should be `getOutputFieldDelimiter`

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 		if reqColNames[0] == "0" {
 			return "", ErrInvalidColumnIndex
 		}
-		// Subtract 1 because AWS Indexing is not 0 based, it starts at 1.
```
Why was the subtraction removed?

sinhaashish replied:
```myRow = writeRow(myRow, jsonValue(string("_"+strconv.Itoa(mytempindex-1)), record), inputType.getOutputFieldDelimitter(), len(reqColNames)) ``` . Substraction is still present. The key to map[string]interfaces{} is being created like "_0" , "_1.

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 
 // processColNameIndex is the function which creates the row for an index based
 // query.
-func (reader *Input) processColNameIndex(record []string, reqColNames []string, columns []string) (string, error) {
```
Why is the record changed from a slice to a string? Was it wrong before?

sinhaashish replied:
We dont get the record as slice now because the code is made generics, We have map[string]interface{} which is decoded to string and thus the record is a string now.

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
-			myVal := evaluateFuncExpr(myFunc.funcExpr[i], "", record, columnsMap)
-			myRow = writeRow(myRow, myVal, reader.options.OutputFieldDelimiter, len(reqColNames))
+			myVal := evaluateFuncExpr(myFunc.funcExpr[i], "", record)
+			myRow = writeRow(myRow, myVal, inputType.getOutPutFieldDelimitter(), len(reqColNames))
```
typo `OutPut` should be `Output` everywhere

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
-		myRow = writeRow(myRow, record[myTempIndex], reader.options.OutputFieldDelimiter, len(reqColNames))
+		myRow = writeRow(myRow, jsonValue(reqColNames[i], record), inputType.getOutPutFieldDelimitter(), len(reqColNames))
 	}
 	if len(myRow) > 1000000 {
```
Can we add a constant for this number somewhere?

kannappanr replied:
Is this more like, we return a max of 1 million records? Is it the same restriction with AWS?

sinhaashish replied:
>Is this more like, we return a max of 1 million records? Is it the same restriction with AWS?

No . it is the maximum character per record is set to be 1 MB.

kannappanr replied:
This should be in a constant.

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 // methods on the given row, it uses an array defined the the main parsing
 // function to keep track of values.
-func aggregationFunctions(counter int, filtrCount int, myAggVals []float64, columnsMap map[string]int, storeReqCols []string, storeFunctions []string, record []string) error {
+
```
Please remove unwanted space here

sinhaashish replied:
Done

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -414,3 +454,25 @@ func aggregationFunctions(counter int, filtrCount int, myAggVals []float64, colu
 }
+
+// convertToSlice takes the map[string]interface{} and convert it to []string
+func convertToSlice(columnsMap map[string]int, record map[string]interface{}, marshshelledRecord string) []string {
```
typo `marshshelledRecord` should be `marshalledRecord`

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
+		individualRecord, err = inputType.Read()
+	case *JSONInput:
+		individualRecord, err = inputType.Read()
+	}
```
should we return an error if the type doesn't match one of these two?

sinhaashish replied:
There is no specific error in the case you mentioned.

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
+
+}
+
+//ReadIndividualRecord return a map[string]interfaces{} for each row in csv and each json for JSON document
```
Comment text needs to have a space

sinhaashish replied:
done

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 					}
 				}
 			}
+
```
Please remove extra unwanted space

sinhaashish replied:
removed

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -220,25 +231,35 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
+				switch inputType.(type) {
+				case *CSVInput:
+					rowStruct := &Row{
+						record: printAsterix(convertToSlice(columnsMap, individualRecord, string(out)), inputType) + "\n",
```
name can be different, something like `printAllColumns` instead of `printAsterix`

sinhaashish replied:
Changed

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -147,24 +150,21 @@ func (reader *Input) ParseSelect(sqlInput string) ([]string, string, int64, inte
-// the where clause it currently prints the appropriate row given the requested
-// columns.
-func (reader *Input) processSelectReq(reqColNames []string, alias string, whereClause interface{}, limitOfRecords int64, functionNames []string, myRow chan *Row, myFunc *SelectFuncs) {
+func processSelectReq(reqColNames []string, alias string, whereClause interface{}, limitOfRecords int64, functionNames []string, myRow chan *Row, myFunc *SelectFuncs, inputType SelectQuery) {
```
Can you add a comment about this function?

sinhaashish replied:
Comment added

kannappanr commented on pkg/s3select/json-input.go:
```diff
@@ -0,0 +1,143 @@
+// If there is a problem with reading the first Json, the error is returned.
+// Otherwise, the returned reader can be reliably consumed with jsonRead()
+// until jsonRead() returns nil.
+func NewJSONInput(opts1 *JSONOptions) (*JSONInput, error) {
```
`opts1` can be `opts`

sinhaashish replied:
Done

kannappanr commented on pkg/s3select/json-input.go:
```diff
@@ -0,0 +1,143 @@
+	myReader := opts1.ReadFrom
+	var tempBytesScanned int64
+	progress := &statInfo{
+		BytesScanned:   tempBytesScanned,
```
`BytesScanned:   tempBytesScanned,` can be changed to `BytesScanned:   int64(0),` or just `BytesScanned:   0,`

sinhaashish replied:
Didn't modify

kannappanr replied:
I see that the code has changed and this comment is not relevant anymore.

kannappanr commented on pkg/s3select/input.go:
```diff
@@ -116,95 +153,34 @@ type Options struct {
-			row = append(row, "")
-		}
-	}
+	//To be implemented
```
name `Type` is not very intuitive, maybe this can be removed and added later, when you have an implementation ready.

sinhaashish replied:
Done

kannappanr commented on pkg/s3select/input.go:
```diff
@@ -116,95 +153,34 @@ type Options struct {
-	// QuoteCharacter - " (defaulted currently)
-	reader.reader.LazyQuotes = true
+	// ReadFrom is where the data will be read from.
+	//ReadFrom io.Reader
```
Remove this commented code

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/helpers.go:
```diff
@@ -193,6 +190,138 @@ func applyStrFunc(rawArg string, funcName string) string {
+		switch colToMyVal := colTo.(type) {
+		case *sqlparser.FuncExpr:
+			var temp string
+			temp = stringOps(colToMyVal, record, "")
```
why is this converted to a byte array here and not in the following piece of code?

sinhaashish replied:
This was the same implementation as on master branch, i have removed it now.

kannappanr commented on pkg/s3select/helpers.go:
```diff
@@ -193,6 +190,138 @@ func applyStrFunc(rawArg string, funcName string) string {
+		}
+	}
+	var myFuncVal string
+	// myFuncVal = ""
```
Please remove the commented code.

sinhaashish replied:
Done

kannappanr commented on pkg/s3select/helpers.go:
```diff
@@ -193,6 +190,138 @@ func applyStrFunc(rawArg string, funcName string) string {
+
+// evalBetweenGreater is a function which evaluates the between given that the
+// TO is > than the FROM.
+func evalBetweenGreater(conversionColumn string, record string, funcName string, colFromVal interface{}, colToVal interface{}, myCoalVal string) (bool, error) {
```
should `myCoalVal` be `myColVal`? or is it the short form for `coalesced` value?

sinhaashish replied:
It it the value from column value . Changed `myCoalVal` to `myColVal`

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/helpers.go:
```diff
@@ -70,33 +73,34 @@ func representsInt(s string) bool {
 		}
 		if operator == "not between" {
-			myResult, err := evaluateBetween(expr, alias, row, columnNames)
+			myResult, err := evaluateBetween(expr, alias, string(out))
```
Can you rename the variable to something like `result` instead of `myResult`?

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/funcEval.go:
```diff
@@ -140,26 +138,54 @@ func evaluateFuncExpr(myVal *sqlparser.FuncExpr, myReturnVal string, myRecord []
 
 // evaluateFuncErr is a function that flags errors in nested functions.
-func (reader *Input) evaluateFuncErr(myVal *sqlparser.FuncExpr) error {
+func (reader *CSVInput) evaluateFuncErr(myVal *sqlparser.FuncExpr) error {
```
this function and the one following are both doing the same thing. Can the function be changed to `func evaluateFuncErr(myVal *sqlparser.FuncExpr, reader *SelectQuery) error`

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/funcEval.go:
```diff
@@ -117,9 +117,7 @@ func processCoalNoIndex(coalStore []string) string {
 }
 
-// evaluateFuncExpr is a function that allows for tail recursive evaluation of
-// nested function expressions.
```
The comments can be added back.

sinhaashish replied:
Done.

kannappanr commented on cmd/object-handlers.go:
```diff
@@ -196,26 +195,43 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
+
+	if selectReq.InputSerialization.CSV != nil {
+
+		if selectReq.InputSerialization.CSV.FileHeaderInfo != s3select.CSVFileHeaderInfoUse &&
```
We should probably add mint tests to test these scenarios.

sinhaashish replied:
Will do in another PR

kannappanr replied:
Yes. Mint tests are not part of minio. What I meant was we should create issues and track it.

poornas committed:
Fix: Disallow requests with SSE-KMS headers (#6587)

Addresses issue #6582. Minio server currently does not
have SSE-KMS support. Reject requests with SSE-KMS headers
with NotImplementedErr

kannappanr commented on pkg/s3select/csv-input.go:
```diff
@@ -0,0 +1,224 @@
+	return nil, nil
+}
+
+// It returns the delimitter specified in input request
```
typo `delimitter` should be `delimiter`

sinhaashish replied:
Done

kannappanr commented on pkg/s3select/csv-input.go:
```diff
@@ -0,0 +1,224 @@
+	return reader.header
+}
+
+// ReadRecord reads a single record from the . Always returns successfully.
```
This comment doesn't make sense

sinhaashish replied:
Done.

kannappanr commented on pkg/s3select/csv-input.go:
```diff
@@ -0,0 +1,224 @@
+	return xml.Header + string(out), nil
+}
+
+// createProgressXML is the function which does the marshaling from the progress structs into XML so that the progress and stat message can be sent
```
can this comment be written in two lines?

sinhaashish replied:
Done

sinhaashish replied:
Written in two lines

kannappanr commented on pkg/s3select/select.go:
```diff
@@ -282,75 +309,97 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
-// to keep track of values.
-func aggregationFunctions(counter int, filtrCount int, myAggVals []float64, columnsMap map[string]int, storeReqCols []string, storeFunctions []string, record []string) error {
+// aggregationFunctions is a function which performs the actual aggregation
+// methods on the given row, it uses an array defined the the main parsing
```
Should `it uses an array defined the the ` be `it uses an array defined in the `

sinhaashish replied:
Done

sinhaashish replied:
Modified

harshavardhana commented on cmd/object-handlers.go:
```diff
@@ -199,26 +198,43 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
+		if selectReq.InputSerialization.JSON.Type != s3select.JSONTypeDocument &&
+			selectReq.InputSerialization.JSON.Type != s3select.JSONLinesType &&
+			selectReq.InputSerialization.JSON.Type != "" {
+			writeErrorResponse(w, ErrInvalidJSONType, r.URL)
```
We have to return here we are proceeding the handler...

sinhaashish replied:
Done

harshavardhana commented on pkg/s3select/input.go:
```diff
@@ -288,20 +218,31 @@ type Row struct {
+
+	case *CSVInput:
+		expression = inputType.getExpression()
+	}
```
Unknown input type lets return an error right here..

sinhaashish replied:
Its already checked in object-handler 
```if selectReq.InputSerialization.CompressionType == "" {
		selectReq.InputSerialization.CompressionType = s3select.SelectCompressionNONE
		if !strings.Contains(objInfo.ContentType, "text/csv") && !strings.Contains(objInfo.ContentType, "application/json") {
			writeErrorResponse(w, ErrInvalidDataSource, r.URL)
			return
		}
	}
```

Then Why do we need the check here again?

harshavardhana replied:
Yes but since this is an Init for s3select package we can return an error.

harshavardhana commented on pkg/s3select/input.go:
```diff
@@ -314,7 +255,8 @@ func (reader *Input) Execute(writer io.Writer) error {
 				return nil
 			} else if ok {
-				message := reader.writeRecordMessage(row.record, curBuf)
+				jsonString := row.record
```
any reason to keep this variable? why not use `row.record` ?

sinhaashish replied:
Removed

harshavardhana commented on pkg/s3select/select.go:
```diff
@@ -175,41 +178,56 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
+			myRow <- rowStruct
+			return
+		}
+		out, _ := json.Marshal(individualRecord)
```
what if the record was CSV?

sinhaashish replied:
CSV anf JSON bot are read as map[string]interface so `out, _ := json.Marshal(individualRecord)` is valid for both.

harshavardhana replied:
I am not able to understand how json.Marshal will be the same as CSV, it is not easy to comprehend this tyle of logic.

harshavardhana commented on pkg/s3select/select.go:
```diff
@@ -175,41 +178,56 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
+		}
+		out, _ := json.Marshal(individualRecord)
+		inputType.getBytesProcessed(individualRecord)
+		if individualRecord == nil {
```
this style is not clean, we shouldn't use `nil` to indicate that its CSV

sinhaashish replied:
@harshavardhana  : Can you please suggest a better approach. Will incorporate it.

harshavardhana replied:
The problems with this approach are few basically json.Marshal errors will be ignored, we are combining both logic i.e json and CSV which are two different types of data into one framework. 

CSV cannot be parsed as JSON but there is a new overhead now to trying to marshal it anyways while failing to do so we send a CSV message.

The flow should be based on inputType, where we process these separately. i.e CSV shouldn't be processed as JSON at all.

harshavardhana commented on pkg/s3select/select.go:
```diff
@@ -220,25 +238,34 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 			if reqColNames[0] == "*" && functionNames[0] == "" {
-				rowStruct := &Row{
-					record: reader.printAsterix(record) + "\n",
+				switch inputType.(type) {
```
This can be written as 

```
var row *Row

switch inputType.(type) {
case *CSVInput:
    row = &Row{}
case *JSONInput:
  row = &Row{}
}
myRow <- row
```

sinhaashish replied:
Done

harshavardhana commented on pkg/s3select/select.go:
```diff
@@ -282,75 +309,97 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
-		}
-		row[i] = record[myTempIndex]
+		row[i] = jsonValue(reqColNames[i], record)
+		//myRow = writeRow(myRow, jsonValue(reqColNames[i], record), inputType.getOutputFieldDelimiter(), len(reqColNames))
```
Why is this commented?

sinhaashish replied:
Done

harshavardhana commented on pkg/s3select/select.go:
```diff
@@ -405,3 +452,25 @@ func aggregationFunctions(counter int, filtrCount int, myAggVals []float64, colu
 }
+
+// convertToSlice takes the map[string]interface{} and convert it to []string
+func convertToSlice(columnsMap map[string]int, record map[string]interface{}, marshalledRecord string) []string {
```
```
convertToSlice()
```
why is this function needed?

sinhaashish replied:
The input type is whether JSON/CSV is read as map[string]interface. Now when the output is csv , the marshalled record is converted to  slice in the particular order , which is read from `columnsMap map[string]int` which keeps the sequencing of column intact.

harshavardhana commented on pkg/s3select/json-input.go:
```diff
@@ -0,0 +1,156 @@
+
+// Populates the bytes Processed
+func (reader *JSONInput) getBytesProcessed(individualRecord map[string]interface{}) {
+	out, _ := json.Marshal(individualRecord)
```
this error cannot be ignored... we should either panic here.

harshavardhana reviewed (changes requested):
your commit ID shows commit from @poornas - can you fix your commit message and make sure that the rebase is on top of the latest master.

Better yet close this PR and re-open another one with proper commit message etc.
